### PR TITLE
fix(bbox): reject rank-4 input in infer_bbox_shape3d and bbox_to_mask3d (closes #4248)

### LINKED ISSUE #4248: Bug: `infer_bbox_shape3d` / `bbox_to_mask3d` accept `(B, N, 8, 3)` through `validate_bbox3d` and then crash or return wrong-shaped values
`validate_bbox3d` documents and accepts rank-4 `(B, N, 8, 3)` input (it reshapes to `(B*N, 8, 3)` internally), but its two callers in the same module do not reshape: `infer_bbox_shape3d` runs `torch.index_select(boxes, 1, ...)` and `bbox_to_mask3d` indexes `boxes[:, 4, 2]` on the original tensor, so the box axis is read as the vertex axis. This is the 3D twin of #4180, which #4218 fixed for `infer_bbox_shape` / `bbox_to_mask` by rejecting rank 4 with `ShapeError`.

Reproduced 2026-09-05 at `1eea5835` (main after #4245), PyTorch 2.14.0, Python 3.11, CPU and MPS.

## Reproduction

Run from the checkout root:

```python
import torch
from kornia.geometry.bbox import validate_bbox3d, infer_bbox_shape3d, bbox_to_mask3d

cuboid = torch.tensor([[[1., 1., 1.], [3., 1., 1.], [3., 3., 1.], [1., 3., 1.],
                        [1., 1., 2.], [3., 1., 2.], [3., 3., 2.], [1., 3., 2.]]])
for n in (1, 3, 8):
    boxes = cuboid.expand(n, 8, 3)[None].contiguous()  # (1, n, 8, 3)
    print(n, "validate:", validate_bbox3d(boxes))
    for name, call in (("infer", lambda: infer_bbox_shape3d(boxes)), ("mask", lambda: bbox_to_mask3d(boxes, (4, 5, 5)))):
        try:
            out = call()
            print("  ", name, [tuple(t.shape) for t in out] if isinstance(out, tuple) else tuple(out.shape))
        except Exception as error:
            print("  ", name, type(error).__name__, str(error)[:70])
```

## Observed

```text
1 validate: True
   infer RuntimeError INDICES element is out of DATA bounds, id=1 axis_dim=1
   mask IndexError index 4 is out of bounds for dimension 1 with size 1
3 validate: True
   infer RuntimeError INDICES element is out of DATA bounds, id=5 axis_dim=3
   mask IndexError index 4 is out of bounds for dimension 1 with size 3
8 validate: True
   infer [(1, 3), (1, 3), (1, 3)]
   mask RuntimeError The size of tensor a (4) must match the size of tensor b (3) at non-singleton dimension 3
```

With `N = 8` the vertex-axis reads succeed by coincidence and `infer_bbox_shape3d` returns three `(1, 3)` tensors, one value per coordinate instead of one per box; eight distinct boxes give `depths = [5, 5, 5]` where the flattened `(8, 8, 3)` input gives eight 2s. On MPS the `N = 1` case raises `torch.AcceleratorError` instead of `RuntimeError`; the outcome is the same. `N = 5, 7, 9` were also swept: all raise.

## Expected

Parity with the 2D helpers after #4218: reject rank 4 with `ShapeError` naming the documented `(B, 8, 3)` contract, or flatten it the way `validate_bbox3d` does. Either is a focused fix; the batch-4c convention blocks document the current behavior and carry a wart pin (`test_wart_rank4_input_passes_validate_bbox3d_then_breaks_the_3d_helpers_4248`, `N` in `{1, 8}`) plus a strict xfail asserting the `ShapeError` outcome (`test_convention_rank4_input_is_rejected_by_the_3d_helpers_4248`), so a fix that flattens instead must update that xfail. `kornia/geometry/transform/crop3d.py` calls both helpers with rank-3 input and is unaffected.

## Classification

Broken output: a shape the module's own validator admits crashes with an unrelated indexing error, and for one batch size silently returns wrong-shaped values. Idiom 3 (the accepted shape contract is wider than the implementation's indexing assumptions).

## Disposition

n/a — single symbol

Found during the bbox/boxes conventions audit batch 4c (#4020); the audit changes no runtime behavior.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).

### PR DESCRIPTION
**Pull request by dchaudhari7177**
Closes #4248.

## The bug

`validate_bbox3d` documents and accepts rank-4 `(B, N, 8, 3)`, reshaping to `(B*N, 8, 3)` internally. Neither caller reshapes, and both index dim 1 as the vertex axis:

```python
# infer_bbox_shape3d
torch.index_select(boxes, 1, torch.tensor([1, 2, 5, 6]))
# bbox_to_mask3d
boxes[:, 4, 2]
```

so the box axis is read as the vertices. On main:

```
n=1  validate: True   infer -> IndexError      mask -> IndexError
n=3  validate: True   infer -> IndexError      mask -> IndexError
n=8  validate: True   infer -> three (1, 3) tensors, one value per coordinate
                      mask  -> RuntimeError
```

The eight-box case is the bad one: it returns rather than raising.

## The fix

`KORNIA_CHECK_SHAPE(boxes, ["N", "8", "3"])` / `["B", "8", "3"]` at the top of each caller, which is what #4218 did for the 2D twin of this (#4180).

**Added, not substituted.** #4218 could drop `validate_bbox` because its boolean was discarded and it never raised. `validate_bbox3d` is not that: it raises when the four edges parallel to an axis disagree, so it stays and the shape gate goes in front of it.

`validate_bbox3d` itself is unchanged and still accepts rank 4, since it reshapes correctly. The docstrings now say where the rejection lives and why.

## Tests

The repository already pinned both halves of this: a wart test asserting the broken behaviour, and a `strict=True` xfail convention test asserting the behaviour it should have. The fix flips both, so they collapse into the one convention test, which keeps the `validate_bbox3d(boxes) is True` assertion since that half of the mismatch is deliberate.

```
tests/geometry/test_bbox.py -k 4248                              2 passed
tests/geometry/test_bbox.py                                     47 passed, 1 xfailed
tests/geometry/test_bbox.py + tests/geometry/test_boxes.py     163 passed, 12 xfailed
pytest --doctest-modules kornia/geometry/bbox.py                  7 passed
```

`uvx ruff@0.16.4 check` and `format` clean. CHANGELOG entry under `### Bug fixes`.

`grep -rnE "#4248|issues/4248" kornia/` is now empty -- the three docstring references described the breakage and have been rewritten to describe the rejection instead, including the one in `boxes.py`.

### CODE REVIEW

**Review by ducha-aiki — COMMENTED** [review]
Review at `9107b38b` — **APPROVE after one CHANGELOG edit**.

The shape gates are in the right layer: `validate_bbox3d` keeps its documented rank-4 behavior, while the two callers that index dimension 1 as the vertex axis now reject rank 4 before that indexing. An independent A/B probe reproduced the old one-box exceptions and the eight-box `(1, 3)` wrong result; both callers now raise `ShapeError` for both cardinalities. The focused file passes 89 tests at float32/float64 (2 expected xfails), the `#4248` source prose is cleared while the test pin remains, all 40 GitHub checks are green, and the head is `MERGEABLE`.

The CHANGELOG entry at line 363 cites only issue #4248. Add the PR number as required by the repository contract: `(#4248, #4351)`. No code changes are otherwise needed.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-5.6-sol).

**Review by ducha-aiki — APPROVED** [review]
Delta re-review at `849e9fcb` — **APPROVED**.

The only remaining ask is resolved: the Bug fixes entry now cites both #4248 and #4351. The code and regression are unchanged in substance, and the focused rank-4 rejection pin passed at float32 and float64 (`4 passed`). The merge-tree against current `main` is clean and all 40 hosted checks are green.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-5.6-sol).
